src/main/java/htsjdk/samtools/util/BlockCompressedIndex.java

```diff
@@ -0,0 +1,248 @@
+            this.uncompressedOffset = uncompressedOffset;
+        }
+
+        /** Returns the compressed offset. */
+        public long getCompressedOffset() {
+            return compressedOffset;
+        }
+
+        /** Returns the uncompressed offset. */
+        public long getUncompressedOffset() {
+            return uncompressedOffset;
+        }
+
+        @Override
+        public String toString() {
+            return String.format("IndexEntry={compressed=%d(0x%x),uncompressed=%d(0x%x)",
+                    compressedOffset, compressedOffset,
+                    uncompressedOffset, uncompressedOffset);
+        }
+
+        @Override
+        public boolean equals(final Object obj) {
+            if (obj == null || !(obj instanceof IndexEntry)) {
+                return false;
+            }
+            final IndexEntry other = (IndexEntry) obj;
+            return compressedOffset == other.compressedOffset
+                    && uncompressedOffset == other.uncompressedOffset;
+        }
+
+        @Override
+        public int hashCode() {
+            return 31 * Long.hashCode(compressedOffset) + Long.hashCode(uncompressedOffset);
+        }
+    }
+
+    private final IndexEntry[] entries;
+
+    // private constructors
+    private BlockCompressedIndex(final IndexEntry[] entries) {
+        this.entries = entries;
+    }
+
+    /**
+     * Gets the number of blocks on the file.
+     *
+     * @return the number of blocks.
+     */
+    public int getNumberOfBlocks() {
+        return entries.length;
+    }
+
+    /**
+     * Gets an unmodifiable list with the index entries.
+     *
+     * @return index entries.
+     */
+    public List<IndexEntry> getIndexEntries() {
+        return Collections.unmodifiableList(Arrays.asList(entries));
+    }
+
+    @Override
+    public boolean equals(final Object obj) {
+        if (obj == null || !(obj instanceof BlockCompressedIndex)) {
+            return false;
+        }
+        return Arrays.equals(this.entries, ((BlockCompressedIndex) obj).entries);
+    }
+
+    @Override
+    public int hashCode() {
+        return Arrays.hashCode(entries);
+    }
+
+    @Override
+    public String toString() {
+        return String.format("BlockCompressedIndex:%s", Arrays.toString(entries));
+    }
+
+    /**
+     * Writes this index into the requested path.
+     *
+     * @param output the output file.
+     *
+     * @throws IOException if an I/O error occurs.
+     */
+    public void writeIndex(final Path output) throws IOException {
+        if (output == null) {
+            throw new IllegalArgumentException("null output path");
+        }
+
+        final ByteBuffer buffer = allocateBuffer(entries.length, true);
+
+        // put the number of entries
+        buffer.putLong(Integer.toUnsignedLong(entries.length));
+
+        for (final IndexEntry entry : entries) {
+            // TODO - check if it is unsigned?
+            buffer.putLong(entry.getCompressedOffset());
```

Comment:
I think we should assert that they're not negative somewhere.

Reply:
Good point. I'll do in next commit.

Reply:
Done.